content/things/rust-to-webassembly.md

```diff
@@ -0,0 +1,637 @@
+
+To address this we can use `wasm-strip`, provided yet again by wabt. It’s a tool that removes everything that is unnecessary, including custom sections. After stripping, we are left with a module of a whopping 116B. Disassembling it shows that the only function in that module is called `add` and executes `(f64.add (local.get 0) (local.get 1))`, which means the Rust compiler was able to emit optimal code. Of course, staying on top of binary size gets more complicated with a growing code base.
+
+### Sneaky bloat
+
+I have seen a couple of complaints online about how big WebAssembly modules create by Rust are that do a seemingly small job. In my experience, there are three reasons why Rust emits surprisingly big WebAssembly binaries:
+
+* Debug build (i.e. forgetting to pass `--release` to `Cargo`)
+* Debug symbols (i.e. forgetting to run `wasm-strip`)
+* String formatting and panics 
+
+We have looked at the first two. Let’s take a closer look at the last one. This innocuous line of code compiles to 18KB of WebAssembly:
+
+```rust
+static PRIMES: &[i32] = &[2, 3, 5, 7, 11, 13, 17, 19, 23];
+
+#[no_mangle]
+fn nth_prime(n: usize) -> i32 {
+    PRIMES[n]
+}
+```
+
+Okay, maybe not so innocuous after all. You might already know what's going on here. 
+
+### Panicking
+
+A quick look at `twiggy` shows that the main contributors to the Wasm module size are functions related to string formatting and panicking (before LTO, there was also code for memory management in there). And that makes sense! The parameter `n` is unsanitized and used to index an array. Rust has no choice but to inject bounds checks. If a bounds check fails, Rust panics.
+
+One way to handle this is to do the bounds checking ourselves. Rust's compiler is really good at proving whether undefined behavior can happen or not.
+
+|||codediff|rust
+  fn nth_prime(n: usize) -> i32 {
++     if n < 0 || n >= PRIMES.len() { return -1; }
+      PRIMES[n]
+  }
+|||
+
+Arguably more idiomatic would be to lean into `Option<T>` APIs to control how the error case should be handled:
+
+|||codediff|rust
+  fn nth_prime(n: usize) -> i32 {
+-     PRIMES[n]
++     *PRIMES.get(n).unwrap_or(&-1)
+  }
+|||
+
+A third way would be to use some of the `unchecked` methods that Rust explicitly provides. These open the door to undefined behavior and as such are `unsafe`, but if you are okay carrying the burden to ensure safety, the gain in performance (or file size) can be significant!
+
+|||codediff|rust
+  fn nth_prime(n: usize) -> i32 {
+-     PRIMES[n]
++     unsafe { *PRIMES.get_unchecked(n) }
+  }
+|||
+
+We can try and stay on top of where we might cause a panic and try to handle those paths manually. However, once we start relying on third-party crates this is less and less likely to succeed, because we can't easily change how the library does its error handling internally.
+
+### LTO
```

Comment:
You might also want to add `codegen-units = 1` somewhere here. TL;DR it reduces intra-crate parallelisation, leading to even slower but sometimes even better optimisations (whether size or perf).

I don't know if it will make any difference on a tiny example like in this article - probably not - but worth at least a shout-out as it helps with sizes on larger projects.

Reply:
Ah, and maybe `opt-level = "s"`? If we go for sizes, seems odd not to mention it here.